Move unstructured conversion into pkg/runtime

smarterclayton opened the pull request:
Scheme conversion should support unstructured conversion natively to
allow going from unstructured to typed and back. It is not a higher
level responsibility to do that conversion because the scheme is the
only one who knows what types it supports.

@liggitt @kubernetes/sig-cli-api-reviews I am going to make Scheme support unstructured in ConvertToVersion and Convert, which means resource.Builder and the CLI can get simpler for all existing use cases where versioned and unstructured need to coexist.

smarterclayton committed:
Move unstructured conversion into pkg/runtime

Scheme conversion should support unstructured conversion natively to
allow going from unstructured to typed and back. It is not a higher
level responsibility to do that conversion because the scheme is the
only one who knows what types it supports.

smarterclayton committed:
Scheme should support Unstructured conversion natively

To reduce the impact of transitioning away from internal, make
Unstructured conversion a natural part of the scheme. Convert and
ConvertToVersion now automatically manage converting unstructured
objects into versioned types as necessary.

Alter the signature of Unstructured to make set possible.

smarterclayton commented:
Second commit makes Scheme support unstructured conversion automatically, meaning that resource.Builder can primarily be Unstructured but switch to versioned or internal as necessary.

smarterclayton commented:
Todo: Needs to guard against internal conversion to unstructured by forcing the kind to move to an external type.

krzyzacy commented:
/retest

smarterclayton committed:
Scheme should provide ObjectTyper for Unstructured objects as well

This will allow us to recognize unstructured objects in the absence of
server side discovery info.

smarterclayton commented:
Also added it so that Scheme supports unstructured.  Will allow the client factory to fall back to the legacy scheme if discovery is not possible, even for unstructured.

smarterclayton commented:
/retest

smarterclayton commented:
/retest

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go:
```diff
@@ -74,13 +71,36 @@ func (u *UnstructuredList) UnstructuredContent() map[string]interface{} {
+		newItems = append(newItems, o)
+	}
+	obj.Items = unstructuredItems
+	obj.Object["items"] = newItems
```
obj.Object could be nil here

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/converter.go:
```diff
@@ -122,8 +139,8 @@ func (c *converterImpl) FromUnstructured(u map[string]interface{}, obj interface
 		}
-		if err == nil && !apiequality.Semantic.DeepEqual(obj, newObj) {
-			glog.Fatalf("FromUnstructured mismatch for %#v, diff: %v", obj, diff.ObjectReflectDiff(obj, newObj))
+		if err == nil && !semantic.DeepEqual(obj, newObj) {
```
should use `c.comparison`?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/converter.go:
```diff
@@ -92,24 +100,33 @@ func parseBool(key string) bool {
+// NewUnstructuredConverter creates an UnstructuredConverter that accepts JSON typed maps and translates them
+// to Go types via reflection. If mismatchDetection is true, the converter will crash and exit if
+// the reflection path does not match a decode and encode to JSON for that type.
+func NewUnstructuredConverter(mismatchDetection bool, comparison ...conversion.Equalities) UnstructuredConverter {
```
doc vararg behavior... if you only honor 0 and 1, fail if > 1 is passed

liggitt replied:
actually, separate constructor for special snowflakes that have their own ideas about equality would be clearer

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/converter_test.go:
```diff
@@ -137,14 +137,14 @@ func doRoundTrip(t *testing.T, item interface{}) {
 	}
 
-	newUnstr, err := conversionunstructured.DefaultConverter.ToUnstructured(item)
+	newUnstr, err := runtime.NewUnstructuredConverter(false).ToUnstructured(item)
```
comment on why we're not using the painstakingly exposed double-check feature

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -217,19 +217,14 @@ func (s *Scheme) AllKnownTypes() map[schema.GroupVersionKind]reflect.Type {
 func (s *Scheme) ObjectKinds(obj Object) ([]schema.GroupVersionKind, bool, error) {
+	// Unstructured objects are always considered to have their declared GVK
+	if _, ok := obj.(Unstructured); ok {
+		return []schema.GroupVersionKind{obj.GetObjectKind().GroupVersionKind()}, false, nil
```
make sure `!obj.GetObjectKind().GroupVersionKind().Empty()`, and error if it is

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -396,10 +391,67 @@ func (s *Scheme) Default(src Object) {
+	uIn, okIn := in.(Unstructured)
+	uOut, okOut := out.(Unstructured)
+	switch {
+	case okIn && okOut:
```
rename to `unstructuredIn`/`unstructuredOut`

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -396,10 +391,67 @@ func (s *Scheme) Default(src Object) {
+	case okIn:
+		// converting an unstructured object to any type is modeled by first converting
+		// the input to a versioned type, then running standard conversions
+		typed, err := s.convertUnstructuredToVersioned(uIn)
```
also needs context/groupversioner to pick preferred target version

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -396,10 +391,67 @@ func (s *Scheme) Default(src Object) {
+		if !ok {
+			return fmt.Errorf("unable to convert the internal object type %T to Unstructured without providing a preferred version to convert to", in)
+		}
+		// since out is not mutated, use unsafe conversion
```
clarify comment (already in an unsafe conversion, out got renamed)

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -439,15 +491,30 @@ func (s *Scheme) UnsafeConvertToVersion(in Object, target GroupVersioner) (Objec
+	var t reflect.Type
+
+	if u, ok := in.(Unstructured); ok {
+		typed, err := s.convertUnstructuredToVersioned(u)
```
pass target through

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -439,15 +491,30 @@ func (s *Scheme) UnsafeConvertToVersion(in Object, target GroupVersioner) (Objec
+
+		in = typed
+		t = reflect.TypeOf(in).Elem()
+		copy = false
```
comment why we don't need to copy

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -439,15 +491,30 @@ func (s *Scheme) UnsafeConvertToVersion(in Object, target GroupVersioner) (Objec
+		}
+
+		in = typed
+		t = reflect.TypeOf(in).Elem()
```
reference contract of convertUnstructuredToVersioned always returning a pointer

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -501,6 +567,31 @@ func (s *Scheme) convertToVersion(copy bool, in Object, target GroupVersioner) (
 
+// convertUnstructuredToVersioned attempts to convert an unstructured object to a typed
+// object if possible. It will return false if conversion is not possible, or the versioned
+// Go form of the object. Note that this conversion will lose fields.
```
update doc

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -501,6 +567,31 @@ func (s *Scheme) convertToVersion(copy bool, in Object, target GroupVersioner) (
+// convertUnstructuredToVersioned attempts to convert an unstructured object to a typed
+// object if possible. It will return false if conversion is not possible, or the versioned
+// Go form of the object. Note that this conversion will lose fields.
+func (s *Scheme) convertUnstructuredToVersioned(in Unstructured) (Object, error) {
```
better name... this doesn't convert between two different gvk's, it converts unstructured to typed

liggitt commented:
direction lgtm, helps make lossless/lossy decoding and conversion more intuitive

smarterclayton commented:
Updated with review comments

smarterclayton commented:
/retest

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go:
```diff
@@ -73,6 +71,32 @@ func (u *UnstructuredList) UnstructuredContent() map[string]interface{} {
+func (obj *UnstructuredList) SetUnstructuredContent(content map[string]interface{}) {
+	obj.Object = content
+	if content == nil {
+		return
```
set obj.Items to nil in this case

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/testing/types.go:
```diff
@@ -213,3 +216,102 @@ func (obj *MyWeirdCustomEmbeddedVersionKindField) GroupVersionKind() schema.Grou
+	if err != nil {
+		return schema.GroupVersionKind{}
+	}
+	gvk := gv.WithKind(u.Object["kind"].(string))
```
unchecked cast

liggitt commented on staging/src/k8s.io/client-go/discovery/unstructured.go:
```diff
@@ -49,29 +49,16 @@ func NewUnstructuredObjectTyper(groupResources []*APIGroupResources) *Unstructur
 	}
-
-	return []schema.GroupVersionKind{gvk}, false, nil
+	return []schema.GroupVersionKind{obj.GetObjectKind().GroupVersionKind()}, false, nil
```
this could easily be empty... would rather fail early here

smarterclayton commented:
Fixed - adding label as per verbal

smarterclayton committed:
Make edit test use subtests to identify flake

smarterclayton commented:
Turns out apply (and possibly edit) depend on runtime.Unstructured allowing partial GVKs to be returned from ObkectKinds.

smarterclayton committed:
kubectl apply edit-last-applied should fail when version is missing

If a user edits last-applied-configuration and it doesn't have an
apiVersion or kind, the command should error out. Now that we check for
kind and version on the unstructured object typer, the previously
passing test now correctly fails. Add a new explicit failure test and
make the existing test pass.

smarterclayton commented:
@liggitt pushed a fix to the edit tests to handle that we now get an error (correctly) when edit-last-applied is missing a kind and apiVersion.

smarterclayton commented:
/test

smarterclayton commented:
/retest

liggitt commented:
Lgtm

smarterclayton commented:
/retest

smarterclayton commented:
TestPatch flake 
/retest

ash2k commented on staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go:
```diff
@@ -396,10 +399,68 @@ func (s *Scheme) Default(src Object) {
+	case okIn && okOut:
+		// converting unstructured input to an unstructured output is a straight copy - unstructured
+		// is a "smart holder" and the contents are passed by reference between the two objects
+		unstructuredOut.SetUnstructuredContent(unstructuredIn.UnstructuredContent())
```
@smarterclayton `UnstructuredContent()` may mutate the original object (bot list and non-list) so this may lead to data races if `in` is shared among threads. See #55871 where I'm fixing this issue in another place. We need a thread safe read-only method to get unstructured content.
/cc @sttts
